chapter-06/2_common_books/bookworm_internal_test.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"testing"
+)
+
+var (
+	handmaidsTale = Book{Author: "Margaret Atwood", Title: "The Handmaid's Tale"}
+	oryxAndCrake  = Book{Author: "Margaret Atwood", Title: "Oryx and Crake"}
+	theBellJar    = Book{Author: "Sylvia Plath", Title: "The Bell Jar"}
+	janeEyre      = Book{Author: "Charlotte Brontë", Title: "Jane Eyre"}
+	villette      = Book{Author: "Charlotte Brontë", Title: "Villette"}
+	ilPrincipe    = Book{Author: "Niccolò Machiavelli", Title: "Il Principe"}
+)
+
+func TestLoadBookworms(t *testing.T) {
+	tests := map[string]struct {
+		bookwormsFile string
+		want          []Bookworm
+		wantErr       bool
+	}{
+		"file exists": {
+			bookwormsFile: "testdata/bookworms.json",
+			want: []Bookworm{
+				{Name: "Fadi", Books: []Book{handmaidsTale, theBellJar}},
+				{Name: "Peggy", Books: []Book{oryxAndCrake, handmaidsTale, janeEyre}},
+			},
+			wantErr: false,
+		},
+		"file doesn't exist": {
+			bookwormsFile: "testdata/no_file_here.json",
+			want:          nil,
+			wantErr:       true,
+		},
+		"invalid JSON": {
+			bookwormsFile: "testdata/invalid.json",
+			want:          nil,
+			wantErr:       true,
+		},
+	}
+	for name, testCase := range tests {
+		t.Run(name, func(t *testing.T) {
+			got, err := loadBookworms(testCase.bookwormsFile)
+			if err != nil && !testCase.wantErr {
+				t.Fatalf("expected an error %s, got an empty one", err.Error())
+			}
+
+			if err == nil && testCase.wantErr {
+				t.Fatalf("expected no error, got one %s", err.Error())
+			}
+
+			if !equalBookworms(t, got, testCase.want) {
+				t.Fatalf("different result: got %v, expected %v", got, testCase.want)
+			}
+		})
+	}
+}
+
+// equalBookworms is a helper to test the equality of two lists of Bookworms.
+func equalBookworms(t *testing.T, bookworms, target []Bookworm) bool {
+	t.Helper()
```

Comment:
Do we cover this ?

Reply:
Very briefly and we mentioned that we will explain it later in the book.

Reply:
Let's not put it in the chapter at all.